# Fundacion YPF - Formación en Ciencia de Datos

## Tarea: preparación de un dataset de ventas

En esta actividad vas a aplicar los conceptos trabajados en clase sobre carga, inspección, calidad y limpieza de datos. El archivo a utilizar es `ventas_tarea.csv`.

Este cuaderno contiene únicamente consignas y guías. Creá debajo de cada apartado las celdas de código que necesites, ejecutalas en orden y dejá visibles tus resultados.

## Objetivo y entrega

El objetivo es obtener un DataFrame limpio y validado, listo para una exploración posterior. No se evalúa sólo que el código funcione: también importa que cada decisión de limpieza tenga una justificación.

Al entregar, el cuaderno debe incluir:

- las celdas de código utilizadas;
- una breve explicación Markdown de los hallazgos y decisiones;
- un DataFrame original sin modificar y una copia limpia;
- controles finales que demuestren que la limpieza funcionó.

## Reglas de trabajo

1. Conservá el DataFrame original tal como fue leído.
2. Realizá los cambios sobre una copia, por ejemplo `df_limpio`.
3. Antes de modificar una columna, inspeccionala y explicá qué problema encontraste.
4. No elimines filas, nulos o duplicados sin justificar por qué esa decisión es válida.
5. Ejecutá las celdas de arriba hacia abajo antes de entregar.

## Parte 1 — Carga e inspección inicial

1. Importá pandas y cargá `ventas_tarea.csv` en un DataFrame llamado `df`.
2. Mostrá las primeras y las últimas filas. Describí qué representa cada fila.
3. Determiná la cantidad de filas y columnas.
4. Listá los nombres de las columnas y clasificálas como identificador, texto, categoría, medida numérica o fecha.
5. Revisá los tipos inferidos por pandas y el resumen de valores no nulos.
6. Escribí una observación: ¿qué columna o columnas requerirán atención antes de analizarlas? ¿Por qué?

**Pista:** los métodos `head`, `tail`, `shape`, `columns`, `dtypes` e `info` responden estas preguntas. Usá una celda por consulta para que el razonamiento quede claro.

In [ ]:
import pandas as pd

url = "https://raw.githubusercontent.com/gabykap29/tp_pandas_fundacion_ypf/main/ventas_tarea.csv"

df = pd.read_csv(url)

In [ ]:
df.head()

,id_venta,fecha,nombre,apellido,producto,categoria,cantidad,precio_unitario,total,provincia
0,1,2026-02-22,JUAN,torres,Notebook,Computacion,3,900000,2700000.0,Corrientes
1,2,12-12-2025,Lucia,GOMEZ,Mouse,Perifericos,1,18000,18000.0,Formosa
2,3,20-11-2025,Maria,perez,Auriculares,Audio,4,50000,NaN,Chaco
3,4,17-03-2026,CARLOS,GOMEZ,Monitor,Pantallas,1,280000,280000.0,Corrientes
4,5,24-06-2026,Ana,romero,Auriculares,Audio,4,50000,200000.0,Cordoba


In [ ]:
df.tail()

,id_venta,fecha,nombre,apellido,producto,categoria,cantidad,precio_unitario,total,provincia
1195,1196,2025-01-13,Mateo,TORRES,Notebook,Computacion,5,900000,4500000.0,Chaco
1196,1197,05-03-2025,valentina,SANCHEZ,Webcam,Video,2,70000,140000.0,Corrientes
1197,1198,2025-08-24,Lucia,Torres,Notebook,Computacion,3,900000,NaN,Corrientes
1198,1199,2025-06-03,ana,Fernandez,Impresora,Impresion,5,220000,1100000.0,Buenos Aires
1199,1200,2026-02-05,JUAN,RODRIGUEZ,Impresora,Impresion,2,220000,440000.0,Corrientes


Cada fila representa una venta realizada. Cada registro contiene información sobre la venta, como su identificador, fecha, nombre y apellido del cliente, producto, categoría, cantidad, precio unitario, total y provincia.

In [ ]:
df.shape

(1200, 10)

In [ ]:
df.columns

Index(['id_venta', 'fecha', 'nombre', 'apellido', 'producto', 'categoria',
       'cantidad', 'precio_unitario', 'total', 'provincia'],
      dtype='object')

id_venta es un identificador. fecha corresponde a una fecha. nombre, apellido y producto son datos de texto. categoria y provincia son variables categóricas. cantidad, precio_unitario y total son medidas numéricas.

In [ ]:
df.dtypes

,0
id_venta,int64
fecha,object
nombre,object
apellido,object
producto,object
categoria,object
cantidad,int64
precio_unitario,int64
total,float64
provincia,object


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 10 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id_venta         1200 non-null   int64  
 1   fecha            1200 non-null   object 
 2   nombre           1200 non-null   object 
 3   apellido         1200 non-null   object 
 4   producto         1200 non-null   object 
 5   categoria        1200 non-null   object 
 6   cantidad         1200 non-null   int64  
 7   precio_unitario  1200 non-null   int64  
 8   total            1099 non-null   float64
 9   provincia        1200 non-null   object 
dtypes: float64(1), int64(3), object(6)
memory usage: 93.9+ KB


## Parte 2 — Auditoría de calidad

En esta parte todavía no hay que corregir datos. Primero reuní evidencia.

1. Contá los valores faltantes de cada columna.
2. Filtrá y observá algunas filas donde haya valores faltantes. ¿Qué otras columnas de esas filas podrían ayudar a entender o reconstruir el dato?
3. Contá las filas completamente duplicadas. Explicá por qué encontrar duplicados no implica automáticamente borrarlos.
4. Verificá si `id_venta` es único.
5. Usá conteos de frecuencia para explorar `producto`, `categoria`, `provincia`, `nombre` y `apellido`. Buscá diferencias de escritura, espacios o capitalización.
6. Revisá si las relaciones entre producto y categoría parecen coherentes.

**Concepto:** la auditoría revisa completitud (faltantes), unicidad (repetidos), consistencia (relaciones entre columnas) y validez (formatos o valores aceptables). El resultado de una auditoría es una lista de hallazgos, no una limpieza automática.

In [ ]:
df.isnull().sum()

,0
id_venta,0
fecha,0
nombre,0
apellido,0
producto,0
categoria,0
cantidad,0
precio_unitario,0
total,101
provincia,0


In [ ]:
df[df.isnull().any(axis=1)]

,id_venta,fecha,nombre,apellido,producto,categoria,cantidad,precio_unitario,total,provincia
2,3,20-11-2025,Maria,perez,Auriculares,Audio,4,50000,NaN,Chaco
12,13,2025-11-27,Sofia,TORRES,Webcam,Video,4,70000,NaN,Misiones
44,45,2025-06-05,Maria,GOMEZ,Monitor,Pantallas,4,280000,NaN,Misiones
86,87,29-07-2025,ana,Rodriguez,Auriculares,Audio,3,50000,NaN,Corrientes
104,105,2026-04-22,MARIA,TORRES,Mouse,Perifericos,4,18000,NaN,Formosa
...,...,...,...,...,...,...,...,...,...,...
1177,1178,2025-12-13,PEDRO,SANCHEZ,Impresora,Impresion,2,220000,NaN,Corrientes
1183,1184,03-08-2026,pedro,diaz,Mouse,Perifericos,3,18000,NaN,Formosa
1188,1189,2025-02-11,carlos,ROMERO,Teclado,Perifericos,2,35000,NaN,Formosa
1193,1194,16-03-2025,MATEO,sanchez,Notebook,Computacion,2,900000,NaN,Chaco


Las filas con valores faltantes deben analizarse considerando el resto de sus columnas. En particular, si falta el valor de total, las columnas cantidad y precio_unitario pueden ayudar a reconstruirlo mediante la multiplicación de ambas. No se realizan modificaciones en esta etapa porque primero se busca identificar y comprender los problemas.

In [ ]:
df.duplicated().sum()

np.int64(0)

Encontrar filas duplicadas no implica eliminarlas automáticamente, ya que podrían representar registros válidos o ventas diferentes que casualmente contienen los mismos datos. Antes de eliminarlas sería necesario determinar si realmente corresponden a registros repetidos por error.

In [ ]:
df["id_venta"].is_unique

True

In [ ]:
df["producto"].value_counts()

,count
producto,
Monitor,187
Teclado,181
Notebook,181
Impresora,175
Mouse,174
Webcam,157
Auriculares,145


In [ ]:
df["categoria"].value_counts()

,count
categoria,
Perifericos,355
Pantallas,187
Computacion,181
Impresion,175
Video,157
Audio,145


In [ ]:
df["provincia"].value_counts()

,count
provincia,
Misiones,221
Formosa,205
Corrientes,198
Buenos Aires,196
Chaco,194
Cordoba,186


In [ ]:
df["nombre"].value_counts()

,count
nombre,
lucia,61
CARLOS,51
Sofia,48
Maria,47
valentina,46
PEDRO,45
ana,44
SOFIA,43
Lucia,42


In [ ]:
df["apellido"].value_counts()

,count
apellido,
gomez,53
GOMEZ,52
LOPEZ,48
romero,48
diaz,48
RODRIGUEZ,47
Lopez,44
torres,44
Perez,44


In [ ]:
df[["producto", "categoria"]].drop_duplicates().sort_values("producto")

,producto,categoria
2,Auriculares,Audio
15,Impresora,Impresion
3,Monitor,Pantallas
1,Mouse,Perifericos
0,Notebook,Computacion
5,Teclado,Perifericos
12,Webcam,Video


Las relaciones entre producto y categoría son coherentes. Cada producto aparece asociado a una única categoría y no se observan inconsistencias en las combinaciones analizadas. Por ejemplo, Mouse y Teclado pertenecen a Periféricos, mientras que Notebook pertenece a Computación y Webcam a Video.

## Parte 3 — Diagnóstico de la columna `fecha`

La columna `fecha` debe poder ordenarse y agruparse por períodos. Antes de convertirla, investigá cómo está escrita.

1. Mostrá valores de fecha y sus frecuencias.
2. Verificá si todos tienen el mismo formato. Usá coincidencias de patrones para distinguir fechas con año al comienzo de fechas con año al final.
3. Contá cuántas filas corresponden a cada patrón y confirmá si entre ambos explican todos los registros.
4. Explicá por qué una fecha como `03-01-2026` puede ser ambigua si no conocemos el formato de origen.

**Concepto:** una fecha como texto no es todavía una fecha para pandas. Convertirla sin conocer la convención puede cambiar silenciosamente el significado de día y mes.

In [ ]:
df["fecha"].value_counts()

,count
fecha,
2025-11-14,7
2025-12-01,6
2025-01-15,5
03-03-2026,5
24-09-2025,5
...,...
2025-07-14,1
2025-02-15,1
2026-05-05,1


In [ ]:
patron_ymd = df["fecha"].str.match(r"^\d{4}-\d{2}-\d{2}$", na=False)

patron_ymd

,fecha
0,True
1,False
2,False
3,False
4,False
...,...
1195,True
1196,False
1197,True
1198,True


In [ ]:
patron_dmy = df["fecha"].str.match(r"^\d{2}-\d{2}-\d{4}$", na=False)

patron_dmy

,fecha
0,False
1,True
2,True
3,True
4,True
...,...
1195,False
1196,True
1197,False
1198,False


In [ ]:
patron_ymd.sum()

np.int64(599)

In [ ]:
patron_dmy.sum()

np.int64(601)

In [ ]:
patron_ymd.sum() + patron_dmy.sum()

np.int64(1200)

Hay fechas en dos tipos de formato  (yyyy-mm-dd y dd-mm-yyyy).
La fecha 03-01-2026 puede ser ambigua porque, dependiendo del formato utilizado, puede representar el 3 de enero de 2026 si se interpreta como día-mes-año, o el 1 de marzo de 2026 si se interpreta como mes-día-año. Por eso es necesario conocer el formato de origen antes de convertir las fechas.

## Parte 4 — Reparación guiada de fechas

Implementá la reparación sin sobrescribir `df["fecha"]` al comienzo. Trabajá con tres Series temporales:

1. Creá una primera Serie que interprete sólo las fechas con formato año-mes-día. Usá `to_datetime`, declarando el formato y usando `errors="coerce"`.
2. Creá una segunda Serie que interprete sólo las fechas con formato día-mes-año, también con formato explícito y `errors="coerce"`.
3. Revisá los valores faltantes de ambas Series. ¿Por qué cada una deja una parte de las filas como `NaT`?
4. Creá una tercera Serie que combine las dos anteriores mediante `combine_first`.
5. Verificá que la Serie combinada no dejó fechas sin interpretar.
6. Recién entonces creá `df_limpio` como copia de `df` y reemplazá en esa copia la columna `fecha` por la Serie combinada.
7. Confirmá que el tipo final de `fecha` sea un tipo de fecha y hora.

**Justificación:** separar los formatos permite saber exactamente qué regla se aplicó a cada grupo de filas. `errors="coerce"` convierte las conversiones fallidas en `NaT`, lo que facilita detectarlas en vez de ocultarlas.

In [ ]:
fecha_ymd = pd.to_datetime(
    df["fecha"].where(patron_ymd),
    format="%Y-%m-%d",
    errors="coerce"
)

In [ ]:
fecha_dmy = pd.to_datetime(
    df["fecha"].where(patron_dmy),
    format="%d-%m-%Y",
    errors="coerce"
)

In [ ]:
fecha_ymd.isna().sum()

np.int64(601)

In [ ]:
fecha_dmy.isna().sum()

np.int64(599)

Cada Serie deja una parte de los registros como NaT porque cada una fue creada para interpretar únicamente uno de los formatos presentes en la columna fecha. Las fechas que no coinciden con el formato especificado no pueden convertirse y, debido a errors="coerce", se representan como NaT. Esto permite separar los dos formatos sin generar errores.

In [ ]:
fecha_combinada = fecha_ymd.combine_first(fecha_dmy)

In [ ]:
fecha_combinada.isna().sum()

np.int64(0)

In [ ]:
df_limpio = df.copy()

In [ ]:
df_limpio["fecha"] = fecha_combinada

In [ ]:
df_limpio["fecha"].dtype

dtype('<M8[ns]')

## Parte 5 — Tratamiento de `total`

1. En registros donde `total` esté presente, verificá si coincide con la multiplicación de `cantidad` por `precio_unitario`.
2. Contá los casos en los que esa relación no se cumple.
3. Si la relación es consistente, completá únicamente los valores faltantes de `total` usando esa regla. Usá el método `fillna` sobre `df_limpio`.
4. Volvé a contar los faltantes y verificá otra vez la relación entre las tres columnas.

**Justificación:** un importe puede completarse sólo si existe una regla de negocio comprobable. No uses la media o la mediana de ventas para imputar este campo: modificarían la facturación y no respetarían la operación real. Si la regla no fuera consistente, el valor debería conservarse como faltante o consultarse con la fuente.

In [ ]:
total_calculado = df_limpio["cantidad"] * df_limpio["precio_unitario"]

In [ ]:
df_limpio["total"].eq(total_calculado)

,0
0,True
1,True
2,False
3,True
4,True
...,...
1195,True
1196,True
1197,False
1198,True


In [ ]:
(df_limpio["total"].notna() & (df_limpio["total"] != total_calculado)).sum()

np.int64(0)

In [ ]:
df_limpio[
    df_limpio["total"].notna() &
    (df_limpio["total"] != total_calculado)
]

,id_venta,fecha,nombre,apellido,producto,categoria,cantidad,precio_unitario,total,provincia


In [ ]:
df_limpio[df_limpio["total"].isna()]

,id_venta,fecha,nombre,apellido,producto,categoria,cantidad,precio_unitario,total,provincia
2,3,2025-11-20,Maria,perez,Auriculares,Audio,4,50000,NaN,Chaco
12,13,2025-11-27,Sofia,TORRES,Webcam,Video,4,70000,NaN,Misiones
44,45,2025-06-05,Maria,GOMEZ,Monitor,Pantallas,4,280000,NaN,Misiones
86,87,2025-07-29,ana,Rodriguez,Auriculares,Audio,3,50000,NaN,Corrientes
104,105,2026-04-22,MARIA,TORRES,Mouse,Perifericos,4,18000,NaN,Formosa
...,...,...,...,...,...,...,...,...,...,...
1177,1178,2025-12-13,PEDRO,SANCHEZ,Impresora,Impresion,2,220000,NaN,Corrientes
1183,1184,2026-08-03,pedro,diaz,Mouse,Perifericos,3,18000,NaN,Formosa
1188,1189,2025-02-11,carlos,ROMERO,Teclado,Perifericos,2,35000,NaN,Formosa
1193,1194,2025-03-16,MATEO,sanchez,Notebook,Computacion,2,900000,NaN,Chaco


In [ ]:
df_limpio["total"] = df_limpio["total"].fillna(total_calculado)

In [ ]:
df_limpio["total"].isna().sum()

np.int64(0)

In [ ]:
(df_limpio["total"] == df_limpio["cantidad"] * df_limpio["precio_unitario"]).all()

np.True_

Se verificó la consistencia de la columna total comparándola con el resultado de multiplicar cantidad por precio_unitario. Los valores de total que estaban completos se conservaron y los valores faltantes se pudieron reconstruir utilizando esta relación. Finalmente, se comprobó que todos los registros cumplen la condición de que total es igual a cantidad por precio_unitario.

## Parte 6 — Normalización de texto y variables útiles

1. Elegí una estrategia uniforme para `nombre` y `apellido`: remové espacios de los extremos y normalizá mayúsculas/minúsculas. Aplicala en `df_limpio`.
2. Mostrá nuevamente frecuencias de esos campos para comprobar el efecto.
3. A partir de `fecha`, creá una columna `anio` y una columna `mes` adecuada para agrupar cronológicamente.
4. Explicá por qué una columna de mes como período año-mes suele ser más útil que sólo el número del mes cuando hay más de un año.

**Precaución:** normalizar texto mejora la consistencia de filtros y conteos, pero no crea un identificador de cliente. Dos registros con el mismo nombre y apellido podrían ser personas distintas.

In [ ]:
df_limpio["nombre"] = df_limpio["nombre"].str.strip().str.lower()

In [ ]:
df_limpio["apellido"] = df_limpio["apellido"].str.strip().str.lower()

In [ ]:
df_limpio["nombre"].value_counts()

,count
nombre,
lucia,135
sofia,126
maria,125
carlos,123
pedro,119
juan,116
valentina,116
ana,116
mateo,114


In [ ]:
df_limpio["apellido"].value_counts()

,count
apellido,
gomez,132
rodriguez,131
perez,130
lopez,124
diaz,120
torres,118
martinez,116
romero,112
fernandez,112


In [ ]:
df_limpio["anio"] = df_limpio["fecha"].dt.year

In [ ]:
df_limpio["fecha"].dt.month

,fecha
0,2
1,12
2,11
3,3
4,6
...,...
1195,1
1196,3
1197,8
1198,6


In [ ]:
df_limpio["mes"] = df_limpio["fecha"].dt.to_period("M")

In [ ]:
df_limpio[["fecha", "anio", "mes"]].head()

,fecha,anio,mes
0,2026-02-22,2026,2026-02
1,2025-12-12,2025,2025-12
2,2025-11-20,2025,2025-11
3,2026-03-17,2026,2026-03
4,2026-06-24,2026,2026-06


Utilizar un período año-mes es más útil que utilizar solamente el número del mes porque permite diferenciar un mismo mes perteneciente a distintos años. Por ejemplo, enero de 2025 y enero de 2026 no se mezclarían al realizar agrupaciones. De esta manera, se pueden analizar correctamente las ventas y la facturación a lo largo del tiempo.
La normalización de nombres y apellidos mejora la consistencia de los datos, pero no permite identificar de manera única a un cliente, ya que dos personas diferentes podrían tener el mismo nombre y apellido.

## Parte 7 — Validación final

Construí una sección de controles finales sobre `df_limpio`. Debe demostrar, como mínimo:

- qué columnas tienen valores faltantes;
- que `fecha` es un tipo de fecha;
- que `total` respeta la relación con `cantidad` y `precio_unitario`;
- que el identificador de venta sigue siendo único;
- que el número de filas se conserva, salvo que hayas justificado explícitamente una eliminación.

Escribí debajo una conclusión de 3 a 5 líneas: ¿qué problemas tenía el archivo?, ¿cuáles solucionaste?, ¿qué supuestos hiciste?

In [ ]:
df_limpio.isnull().sum()

,0
id_venta,0
fecha,0
nombre,0
apellido,0
producto,0
categoria,0
cantidad,0
precio_unitario,0
total,0
provincia,0


In [ ]:
df_limpio["fecha"].dtype

dtype('<M8[ns]')

In [ ]:
df_limpio["fecha"].isna().sum()

np.int64(0)

In [ ]:
(
    df_limpio["total"]
    == df_limpio["cantidad"] * df_limpio["precio_unitario"]
).all()

np.True_

In [ ]:
df_limpio["id_venta"].is_unique

True

In [ ]:
df.shape

(1200, 10)

In [ ]:
df_limpio.shape

(1200, 12)

## Parte 8 — Primeras preguntas de análisis

Sin realizar todavía gráficos, resolvé al menos cuatro de las siguientes preguntas con `df_limpio`. Antes de cada respuesta, escribí cuál será la métrica, el filtro y la agrupación.

- ¿Cuántas unidades se vendieron en cada mes?
- ¿Cuál fue la facturación por mes?
- ¿Qué producto vendió más unidades? ¿Cuál facturó más?
- ¿Qué categoría representa el mayor porcentaje de facturación?
- ¿Qué provincia tiene más operaciones de venta?
- ¿Cuál es el importe promedio por operación en cada provincia?
- ¿Cómo se comparan las unidades vendidas entre los años presentes?
- Para un mes que elijas, ¿cuál fue la cantidad vendida y la facturación?

**Recordá:** cantidad vendida es la suma de `cantidad`; facturación es la suma de `total`; y operaciones es el conteo de ventas. Son métricas distintas y pueden llevar a conclusiones diferentes.

In [ ]:
df_limpio.groupby("mes")["cantidad"].sum()

,cantidad
mes,
2025-01,202
2025-02,168
2025-03,188
2025-04,196
2025-05,192
2025-06,143
2025-07,184
2025-08,195
2025-09,176


In [ ]:
df_limpio.groupby("mes")["total"].sum()

,total
mes,
2025-01,46603000.0
2025-02,51340000.0
2025-03,51504000.0
2025-04,49381000.0
2025-05,48993000.0
2025-06,28721000.0
2025-07,42632000.0
2025-08,48796000.0
2025-09,42372000.0


In [ ]:
df_limpio.groupby("producto")["cantidad"].sum().sort_values(ascending=False)

,cantidad
producto,
Teclado,582
Monitor,570
Notebook,564
Impresora,552
Mouse,496
Webcam,459
Auriculares,420


In [ ]:
df_limpio.groupby("producto")["total"].sum().sort_values(ascending=False)

,total
producto,
Notebook,507600000.0
Monitor,159600000.0
Impresora,121440000.0
Webcam,32130000.0
Auriculares,21000000.0
Teclado,20370000.0
Mouse,8928000.0


In [ ]:
df_limpio.groupby("provincia")["cantidad"].sum().sort_values(ascending=False)

,cantidad
provincia,
Misiones,671
Corrientes,619
Buenos Aires,609
Formosa,605
Chaco,580
Cordoba,559


In [ ]:
df_limpio.groupby("provincia")["total"].sum().sort_values(ascending=False)

,total
provincia,
Corrientes,183786000.0
Misiones,168995000.0
Buenos Aires,158961000.0
Formosa,136068000.0
Chaco,131861000.0
Cordoba,91397000.0


In [ ]:
df_limpio.groupby("anio")["cantidad"].sum()

,cantidad
anio,
2025,2251
2026,1392


El año con mayor cantidad de unidades vendidas fue 2025

In [ ]:
df_limpio[df_limpio["mes"] == "2026-03"][["cantidad", "total"]].sum()

,0
cantidad,178.0
total,56637000.0


En enero de 2025 se vendieron 202 unidades y se obtuvo una facturación total de $46.603.000.

## Parte Final
Una vez aplicado todas las transformaciones correspondientes y haber respondido las preguntas, es hora de almacenar nuestro df_limpio en un csv.

Utiliza el metodo correspndinete de pandas para exportar tu df_limpio en un .csv llamado `dataset_ventas_limpio.csv`.

In [ ]:
df_limpio.to_csv("dataset_ventas_limpio.csv", index=False)

In [ ]:
import os

os.path.exists("dataset_ventas_limpio.csv")

True

## Lista de verificación antes de entregar

- [ ] El archivo se carga correctamente y el DataFrame original se conserva.
- [ ] Cada hallazgo de calidad tiene una evidencia mostrada en el cuaderno.
- [ ] Las fechas se repararon usando dos conversiones explícitas y una combinación de Series.
- [ ] Las decisiones sobre faltantes y texto están justificadas.
- [ ] `df_limpio` pasó los controles finales.
- [ ] Se respondieron al menos cuatro preguntas de análisis.
- [ ] El cuaderno se ejecutó de principio a fin sin errores.

## Cierre

Preparar datos es parte del análisis, no un paso previo invisible. Un buen cuaderno permite que otra persona entienda qué se observó, qué se cambió y por qué esa decisión es razonable. En la próxima etapa podremos hacer EDA sobre una base más confiable.